File: src/formatting/limits.ts
export function limitLines(lines: string[], maxLines: number): {
  lines: string[];
  omitted: number;
} {
  if (lines.length <= maxLines) {
    return { lines, omitted: 0 };
  }

  const headCount = Math.ceil(maxLines * 0.65);
  const tailCount = Math.max(1, maxLines - headCount);
  const omitted = lines.length - headCount - tailCount;

  return {
    lines: [
      ...lines.slice(0, headCount),
      `... ${omitted} lines omitted ...`,
      ...lines.slice(-tailCount),
    ],
    omitted,
  };
}

File: internal/formatting/limits.go
package formatting

import "fmt"

func LimitLines(lines []string, maxLines int) ([]string, int) {
	if len(lines) <= maxLines {
		return lines, 0
	}
	headCount := (maxLines*65 + 99) / 100
	tailCount := max(1, maxLines-headCount)
	omitted := len(lines) - headCount - tailCount
	limited := make([]string, 0, maxLines+1)
	limited = append(limited, lines[:headCount]...)
	limited = append(limited, fmt.Sprintf("... %d lines omitted ...", omitted))
	limited = append(limited, lines[len(lines)-tailCount:]...)
	return limited, omitted
}

func max(left, right int) int {
	if left > right {
		return left
	}
	return right
}
